### ISSUE TITLE
Use of llama2 or llama3 as baseline?

### ISSUE THREAD

**Issue by pjj** [Haiyang-W/TokenFormer#4]
In your writeup you mention following Karpathy's baseline recipe for training the gpt-2 architecture. Did you also try instead using his (or other) baseline recipes for training and then replacing llama2, or llama3 models?  Is there any reason to suspect that the difference in the FFN Layer construction in those llamaX models will affect your strategy?

**Comment by Haiyang-W** [comment]
Thank you for your suggestion. 

We haven’t tried it yet, but we believe that when model parameter counts are similar, performance should be roughly the same across models. Some non-model-specific tricks, such as ROPE, might help. 

So, as long as the parameter count is aligned, performance is primarily determined by training data and certain hyperparameters. We’ll give it a try later, but I believe that with aligned parameters, the performance should be quite similar.

**Comment by Haiyang-W** [comment]
Or do you have a different perspective? I’d be happy to discuss. It seems that performance is determined solely by the number of parameters and training data? LLaMAX models are also similar to GPT-2 architecture?

**Comment by pjj** [comment]
The llamaX FFN layer has the same compute as the GPT-2 FFN layer, but uses three weight matrices instead of two (with a swiglu activation instead of gelu), so if you use the same Pattention sizes throughout, it may rebalance compute with relatively more compute on the replaced FFN pieces, which changes the compute table details, but may end up in your favor anyways. Rope, rmsnorm, and the order of normalizations are generic modifications and should not have very different effects if applied on top of gpt2 instead. For a clean standalone llamaX, see the Python training code in Karpathy's llama2.c (or the llama3 version in llm.c branches and the official meta repo).

**Comment by Haiyang-W** [comment]
> The llamaX FFN layer has the same compute as the GPT-2 FFN layer, but uses three weight matrices instead of two (with a swiglu activation instead of gelu), so if you use the same Pattention sizes throughout, it may rebalance compute with relatively more compute on the replaced FFN pieces, which changes the compute table details, but may end up in your favor anyways. Rope, rmsnorm, and the order of normalizations are generic modifications and should not have very different effects if applied on top of gpt2 instead. For a clean standalone llamaX, see the Python training code in Karpathy's llama2.c (or the llama3 version in llm.c branches and the official meta repo).

Apologies for the delayed response; I recently went through a serious illness and am only now feeling a bit better. I’ll look into what you mentioned. However, I still believe that model performance mainly depends on parameter size and data, with the specific model design having minimal impact. For LLaMA X's FFN, you can think of it as incorporating an additional gating function to modify the attention score, which likely won’t have a significant impact on the model's performance. As for our transformer baseline, although it follows GPT-2’s structure, it uses ROPE. It’s essentially the same as the Pythia released by GPT-NeoX, except we didn’t use the GPT-J residuals, and we shared vocab and head weights; everything else should be the same.

Thanks for pointing that repo! :) I will have a look in the few days.

**Comment by Maykeye** [comment]
Since repo asks for papers, and the paper about it is not mentioned in the tokenformer paper or this thread, activation functions are considered in 
[GLU Variants Improve Transformer](https://arxiv.org/pdf/2002.05202) which llama paper cites. 

Note that the  paper uses Adafactor optimizer rather than AdamW

It probably at least will give at enough extra oomph for transformers at 300B tokens (ppl 11.63 in paper) to reach tokenformer at 60B (11.60) at 1.4B


(Also FFN layer of tokenformer has [enough parameters](https://github.com/Haiyang-W/TokenFormer/blob/128f2fe308d79353052c7ca4972dbbdfd89c24dd/configs/tokenformer/150M_eval.yml#L10) to use GEGLU itself)

**Comment by kroggen** [comment]
I made a modification on llama2.c to include Pattention modules. You can see the diff [here](https://github.com/kroggen/llama2.c/commit/1794b38454f21ee5d9bbfba0dbc79f092f5a04aa)

But I was not able to make it train faster than the original code, even when using incremental growth

I didn't do many "ablations" though

Any ideas?
